Metric refactor

pradnic opened the pull request:
As requested [here](https://github.com/envoyproxy/ratelimit/pull/237#pullrequestreview-613685473).

In the current version of the system metric production is done ad hoc to comply with specific requirements, Scopes are initialized in service components, Ex: [1](https://github.com/envoyproxy/ratelimit/blob/c03723f3ff39d2c3153587388fd230b3b76324d2/src/service_cmd/runner/runner.go#L106), [2](https://github.com/envoyproxy/ratelimit/blob/c03723f3ff39d2c3153587388fd230b3b76324d2/src/redis/cache_impl.go#L16), [3](https://github.com/envoyproxy/ratelimit/blob/c03723f3ff39d2c3153587388fd230b3b76324d2/src/redis/cache_impl.go#L20); and metrics are directly set by all components as well Ex: [1](https://github.com/envoyproxy/ratelimit/blob/c03723f3ff39d2c3153587388fd230b3b76324d2/src/limiter/base_limiter.go#L50), [2](https://github.com/envoyproxy/ratelimit/blob/c03723f3ff39d2c3153587388fd230b3b76324d2/src/limiter/base_limiter.go#L77). 

This makes adding service-wide configurations to metric generation (such as [Detailed Metrics](https://github.com/envoyproxy/ratelimit/issues/181)) particularly difficult. In order to accommodate this feature, we made the following refactor. 

This PR refactors metric handling in the ratelimit service, namely:
- Adds stats.Manager structure.
- Centralizes metric creation in the stats.Manager.
- Removes, where possible, metric scope handling from the other service's components.

mattklein123 commented:
cc @nezdolik @ysawa0 would either of you be willing to help review this?

nezdolik commented:
@mattklein123 I can help review this Friday

nezdolik commented on src/config/config_impl.go:
```diff
@@ -104,10 +89,8 @@ func newRateLimitConfigError(config RateLimitConfigToLoad, err string) RateLimit
-	config RateLimitConfigToLoad, parentKey string, descriptors []yamlDescriptor,
-	statsScope stats.Scope) {
+// @param manager that owns the stats.Scope.
+func (this *rateLimitDescriptor) loadDescriptors(config RateLimitConfigToLoad, parentKey string, descriptors []yamlDescriptor, manager stats.Manager) {
```
nit: rename arg name manager ->statsManager

nezdolik commented on src/config/config_impl.go:
```diff
@@ -268,13 +235,12 @@ func (this *rateLimitConfigImpl) GetLimit(
 
 	if descriptor.GetLimit() != nil {
-		rateLimitKey := domain + "." + this.descriptorToKey(descriptor)
+		rateLimitKey := stats.DescriptorKey(domain, descriptor)
```
what is the reasoning behind moving `descriptorToKey` function to stats package?

pradnic replied:
Moved to client package (where it is used currently).

nezdolik commented on src/config/config_impl.go:
```diff
@@ -316,9 +282,9 @@ func (this *rateLimitConfigImpl) GetLimit(
 // @return a new config.
 func NewRateLimitConfigImpl(
-	configs []RateLimitConfigToLoad, statsScope stats.Scope) RateLimitConfig {
+	configs []RateLimitConfigToLoad, manager stats.Manager) RateLimitConfig {
```
nit: same as above, rename `manager` to smth like `statsManager`

nezdolik commented on src/config_check_cmd/main.go:
```diff
@@ -19,9 +21,9 @@ func loadConfigs(allConfigs []config.RateLimitConfigToLoad) {
-	dummyStats := stats.NewStore(stats.NewNullSink(), false)
-	config.NewRateLimitConfigImpl(allConfigs, dummyStats)
+	settingStruct := settings.NewSettings()
+	manager := stats.NewStatManager(gostats.NewStore(gostats.NewNullSink(), false), settingStruct)
```
nit: inline creation of settings here instead

nezdolik commented on src/limiter/base_limiter.go:
```diff
@@ -109,25 +111,26 @@ func (this *BaseRateLimiter) GetResponseDescriptorStatus(key string, limitInfo *
 		cacheKeyGenerator:          NewCacheKeyGenerator(cacheKeyPrefix),
 		localCache:                 localCache,
 		nearLimitRatio:             nearLimitRatio,
+		Manager:                    manager,
```
again this field name is quite generic, unless looking at the type, it would be hard to figure out where does this manager belong to

nezdolik commented on src/stats/manager_impl.go:
```diff
@@ -0,0 +1,119 @@
+	WithinLimit             gostats.Counter
+}
+
+func (this RateLimitStats) String() string {
```
this may be just me, but it's not obvious that to get `Key` field from `RateLimitStats` type , one should call `String()` method. I would rather call this method `Key()` or `GetKey()`.

nezdolik commented on test/limiter/base_limiter_test.go:
```diff
@@ -22,10 +23,11 @@ func TestGenerateCacheKeys(t *testing.T) {
 	timeSource := mock_utils.NewMockTimeSource(controller)
 	jitterSource := mock_utils.NewMockJitterRandSource(controller)
 	statsStore := stats.NewStore(stats.NewNullSink(), false)
+	sm := mockstats.NewMockStatManager(statsStore)
```
nit: rename to smth like `statsManager`

nezdolik commented on src/stats/manager.go:
```diff
@@ -0,0 +1,11 @@
+import stats "github.com/lyft/gostats"
+
+type Manager interface {
+	NewStats(key string) RateLimitStats
```
could you add docs to interface methods?

nezdolik commented on src/stats/manager_impl.go:
```diff
@@ -0,0 +1,119 @@
+	logger "github.com/sirupsen/logrus"
+)
+
+func NewStatManager(store gostats.Store, s settings.Settings) *ManagerImpl {
```
renaming `s` settings would probably clash with `settings` package, but it could have improved readability

nezdolik commented on src/stats/manager_impl.go:
```diff
@@ -0,0 +1,119 @@
+}
+
+func (this *ManagerImpl) NewShouldRateLimitStats() ShouldRateLimitStats {
+	s := this.serviceStatsScope.Scope("call.should_rate_limit")
```
nit: s-> scope

nezdolik commented on src/stats/manager_impl.go:
```diff
@@ -0,0 +1,119 @@
+//todo: Ideally the gostats package fields should be unexported
+//	the inner value could be interacted with via getters such as rlStats.TotalHits() uint64
+//	This ensures that setters such as Inc() and Add() can only be managed by ManagerImpl.
+type RateLimitStats struct {
```
could be better to group all those struct definitions together either in top of this file or pull them out into `manager.go` file

nezdolik commented on src/stats/manager_impl.go:
```diff
@@ -0,0 +1,119 @@
+		rlStatsScope:         serviceScope.Scope("rate_limit"),
+		legacyStatsScope:     serviceScope.Scope("call.should_rate_limit_legacy"),
+		serviceStatsScope:    serviceScope,
+		detailedMetricsScope: serviceScope.Scope("rate_limit").Scope("detailed"),
```
should this scope be initialised only if `detailed` is set to true?

pradnic replied:
Removed all mentions of `detailedMetrics` they will be added on the DetailedMetrics PR

nezdolik commented on src/stats/manager_impl.go:
```diff
@@ -0,0 +1,119 @@
+	legacyStatsScope     gostats.Scope
+	serviceStatsScope    gostats.Scope
+	detailedMetricsScope gostats.Scope
+	detailed             bool
```
how is this field currently used?

pradnic replied:
Removed all mentions of `detailedMetrics` they will be added on the DetailedMetrics PR

nezdolik reviewed (commented):
Does extending existing settings with option to include detailed stats come in separate PR?

pradnic commented:
Thank you @nezdolik for the review.

Finished working on the requested changes. Alongside the nit's these are the most important points:

- Renamed `stats.Manager` variables to `statsManager` in order to improve clarity.
- Added documentation on `manager.go` types and methods.
- Moved type definitions from `manager_impl.go` to `manager.go`

nezdolik commented on src/stats/manager.go:
```diff
@@ -0,0 +1,65 @@
+import stats "github.com/lyft/gostats"
+import gostats "github.com/lyft/gostats"
+
+// Manager is the interface that wraps initialization of stat structures
```
nit: add '.' in the end of each doc sentence

nezdolik reviewed (commented):
PR looks good to me. @mattklein123 @dweitzman would you take a look?

pradnic commented:
Hi @mattklein123 let me know if we need any more work on this PR or if we can go forwards with the next steps in order to add detailed-metrics feature.

pradnic commented:
Hi @mattklein123 what are the next steps you suggest in order to get this feature merged in upstream?

mattklein123 commented:
Merge main, and then I will take a look.

/wait

mattklein123 commented:
Please check CI

/wait

pradnic commented on test/service/ratelimit_test.go:
```diff
@@ -139,8 +143,8 @@ func TestService(test *testing.T) {
-		func([]config.RateLimitConfigToLoad, stats.Scope) {
-			barrier.signal()
+		func([]config.RateLimitConfigToLoad, stats.Manager) {
+			defer barrier.signal()
```
@mattklein123 I believe the CI build failed because of a race-condition in this test. The scenario is similar to the one fixed in [this recently merged PR](https://github.com/envoyproxy/ratelimit/pull/258).
I believe the issue should be resolved with this change.

mattklein123 reviewed (approved):
Thanks! Just to make sure I understand, there should be no functional change for this PR, right? In a future PR you will add the ability to have detailed stats?

/wait-any

pradnic commented:
> Thanks! Just to make sure I understand, there should be no functional change for this PR, right? In a future PR you will add the ability to have detailed stats?

Correct! This changeset holds no functional changes. The opt-in Detailed Metrics feature will come in the next PR.

dweitzman commented:
> @dweitzman would you take a look?

I didn't read all the code, but the overall goal makes sense to me. I'm in favor of anything that isolates the `github.com/lyft/gostats` imports so they can more easily be wrapped in or replaced by a more modern, flexible, or expressive way to tracking metrics.
